sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala

```diff
@@ -67,61 +81,78 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
+        (m, mWrapper, false)
       }
 
     case plan: LogicalPlan =>
-      val relations = ArrayBuffer.empty[MultiInstanceRelation]
+      val relations = new mutable.HashSet[ReferenceEqualPlanWrapper]()
+      var planChanged = false
       val newPlan = if (plan.children.nonEmpty) {
-        val newChildren = ArrayBuffer.empty[LogicalPlan]
+        val newChildren = mutable.ArrayBuffer.empty[LogicalPlan]
         for (c <- plan.children) {
-          val (renewed, collected) = renewDuplicatedRelations(existingRelations ++ relations, c)
+          val (renewed, collected, changed) =
+            renewDuplicatedRelations(existingRelations ++ relations, c)
           newChildren += renewed
           relations ++= collected
+          if (changed) {
+            planChanged = true
+          }
         }
 
-        if (plan.childrenResolved) {
-          val attrMap = AttributeMap(
-            plan
-              .children
-              .flatMap(_.output).zip(newChildren.flatMap(_.output))
-              .filter { case (a1, a2) => a1.exprId != a2.exprId }
-          )
-          plan.withNewChildren(newChildren.toSeq).rewriteAttrs(attrMap)
+        if (planChanged) {
+          if (plan.childrenResolved) {
+            val planWithNewChildren = plan.withNewChildren(newChildren.toSeq)
+            val attrMap = AttributeMap(
+              plan
+                .children
+                .flatMap(_.output).zip(newChildren.flatMap(_.output))
+                .filter { case (a1, a2) => a1.exprId != a2.exprId }
+            )
+            if (attrMap.isEmpty) {
+              planWithNewChildren
+            } else {
+              planWithNewChildren.rewriteAttrs(attrMap)
+            }
+          } else {
+            plan.withNewChildren(newChildren.toSeq)
+          }
         } else {
-          plan.withNewChildren(newChildren.toSeq)
+          plan
         }
       } else {
         plan
       }
 
       val planWithNewSubquery = newPlan.transformExpressions {
         case subquery: SubqueryExpression =>
-          val (renewed, collected) = renewDuplicatedRelations(
+          val (renewed, collected, changed) = renewDuplicatedRelations(
             existingRelations ++ relations, subquery.plan)
           relations ++= collected
+          if (changed) planChanged = true
           subquery.withNewPlan(renewed)
       }
-      (planWithNewSubquery, relations.toSeq)
+      (planWithNewSubquery, relations, planChanged)
   }
 
+  @inline
   private def isDuplicated(
-      existingRelations: Seq[MultiInstanceRelation],
-      relation: MultiInstanceRelation): Boolean = {
-    existingRelations.exists { er =>
-      er.asInstanceOf[LogicalPlan].outputSet
-        .intersect(relation.asInstanceOf[LogicalPlan].outputSet).nonEmpty
-    }
+      existingRelations: mutable.HashSet[ReferenceEqualPlanWrapper],
+      planWrapper: ReferenceEqualPlanWrapper): Boolean = {
+    existingRelations.contains(planWrapper)
```

Comment:
So the logic of `isDuplicated` becomes checking whether the plan exists instead of checking outputSet intersection? Is that a safe change?

Reply:
It's safe. For the `MultiInstanceRelation`, it's impossible to have partial duplicate attributes with others as its attributes are always generated by itself, which results in the monotonically increasing exprid.